File: counting-sort/counting-sort.js
function random_sequence(minimum, maximum) { //returns a shuffled array
    array = []
    for (i = minimum; i < maximum; i++) { //create array and append numbers to it
        array.push(i)
    }
    array.sort(() => Math.random() - 0.5) //shuffle the array
    return array
}

function counting_sort(array) {
    //create array with zeros
    let count = []
    for (i = 0; i < Math.max(...array); i++) {
        count.push(0)
    }

    for (i = 0; i < array.length; i++) { //iterate through given array, and add 1 to the index which is the value of array[i]
		count[array[i]]++
	}

	for (i = 1; i < count.length; i++) { //go through array and add previous index's value to the current index
		count[i] += count[i-1]
    }
    
    //create output array filled with null
    let output = []
    for (i = 0; i < array.length; i++) {
        output.push(null)
    }

    //iterate through array and turn none types into sorted values
    array.forEach( i =>
        output[count[i]-1] = i,
        count[i] -= 1
    )

    return output
}

let shuffled_array = random_sequence(0, 1000)
console.log("COUNTING SORT")
console.log("[" + shuffled_array.toString() + "]", "\n")
console.log("[" + counting_sort(shuffled_array).toString() + "]")

File: counting-sort/counting-sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

using std::cout;
using std::endl;

std::vector<int> random_sequence(int minimum, int maximum) { //returns a shuffled array
    std::vector<int> array;
    for (int i = minimum; i < maximum; i++) { //create array and append numbers to it
        array.push_back(i);
    }
    std::shuffle(std::begin(array), std::end(array), std::default_random_engine()); //shuffle the array
    return array;
}

void printArray(std::vector<int> &array) {
    std::cout << "[";
    for (unsigned int i = 0; i < array.size()-1; i++) {
        std::cout << std::to_string(array[i]) << ", ";
    }
    std::cout << std::to_string(array[array.size()-1]) << "]" << std::endl;
}

int max(std::vector<int> &array) { //gets largest number in array
    int largest = array[0];
    for (const int &i : array) {
        if (i > largest) {
            largest = i;
        }
    }
    return largest;
}

std::vector<int> counting_sort(std::vector<int> &array) {
    std::vector<int> count;
    std::vector<int> output; //create output array

    //fill arrays
    for (unsigned int i = 0; i < max(array) + 1; i++) {
        count.push_back(0);
    } for (unsigned int i = 0; i < array.size(); i++) {
        count[array[i]]++;
    }    

    for (unsigned int i = 1; i < count.size(); i++) { //iterate through given array, and add 1 to the index which is the value of array[i]
        count[i] += count[i-1];
    }

    for (unsigned int i = 0; i < array.size(); i++) { //go through array and add previous index's value to the current index
        output.push_back(0);
    }

    for (const int &i : array) { //iterate through array and plug in sorted values
        output[count[i]-1] = i;
        count[i]--;
    }

    return output;
}

int main() {
    cout << "COUNTING SORT" << endl;
    std::vector<int> shuffled_array = random_sequence(0, 1000);

    printArray(shuffled_array);
    cout << endl;

    std::vector<int> sorted_array = counting_sort(shuffled_array);
    printArray(sorted_array);
    return 0;
}